db/view/view.hh

```diff
@@ -80,6 +80,50 @@ struct view_and_base {
     base_info_ptr base;
 };
 
+// An immutable representation of a clustering or static row of the base table.
+struct clustering_or_static_row {
+private:
+    std::optional<clustering_key_prefix> _key;
+    row_tombstone _deleted_at;
+    row_marker _marker;
+    row _cells;
```

Comment:
> You also have the type `deletable_row` which includes exactly the last three items (deleted_at, marker and cells). The "class clustering_row" uses that instead of the three separate items. Please consider if we should use it here too. An example benefit of using deletable row is that your empty() can simply call deletable_row.empty() instead of implementing it from scratch as you did here.

I was a little hesitant to do it because deletable_row is primarily used in clustering rows - and a small part of its code assumes that it only holds regular columns. However, this assumption can easily be removed from the code, so I'll do it and use deletable_row here.

Reply:
> An even more radical idea that you can consider: It seems that the new clustering_or_static_row is basically identical to the existing clustering_row except that the clustering_key_prefix is optional in this variant. But do we need this difference? Can't we use a normal clustering_key_prefix but keep it **empty** and then treat this empty clustering key as a sign that this is a static row?
> 
> I think that if do this a lot of boilerplate wrapping code in this patch can be avoided, and the risk that one day somebody will improve clustering_row but will forget to modify this one will go down.

I'm not entirely a fan of this idea. Unlike deletable_row, to me the name "clustering_row" clearly indicates that it holds data of a clustering row, not of a static row; those are two separate entities. I understand that just using the clustering_row may make it easier to migrate the current logic, but I'm afraid that it may be a source of bugs as other developers may misinterpret a "clustering_row that is actually a static row" as a "true clustering_row".

> 
> If you use my suggest (if it's even poissble...) to use the existing "clustering_row" with an empty clustering_key_prefix, I think that another benefit is that some of the other patches will become smaller - the compute column function doesn't need to change, for example - it can still take just a "clustering_row" (and check if the key is empty if it wishes - but existing implementations don't), and a lot of changes to the types are simply not needed.

I'd rather have a separate clustering_or_static_row type for the sake of type correctness. However, perhaps we can put a clustering_row field in it as its implementation in the way you suggest for the sake of code reuse.

Reply:
> I'd rather have a separate clustering_or_static_row type for the sake of type correctness. However, perhaps we can put a clustering_row field in it as its implementation in the way you suggest for the sake of code reuse.

Good idea!
The only thing that I'm less than happy is that it means that the patch will still contain a gazillion changes of "clustering_row" to "clustering_or_static_row". I'm not sure it's bad - as you said it might force readers to remember this option exists - but on the other hand, **forcing** the reader to think about this possibility when it's not even relevant is weird. I don't know... You can try and see what you think is better, you are much deeper into this code now, and probably understand it and its "spirit" better than I do.

Reply:
I collapsed the `_deleted_at`, `_marker` and `_cells` fields into a single `deletable_row`.